netutil/netutil.go

```diff
@@ -54,3 +57,166 @@ func privateNetworkInterfaces(all []net.Interface, fallback []string, logger log
 	}
 	return privInts
 }
+
+// GetFirstAddressOf returns the first IPv4/IPV6 address of the supplied interface names, omitting any link-local addresses.
+func GetFirstAddressOf(names []string, logger log.Logger, enableInet6 bool) (string, error) {
+	return getFirstAddressOf(names, logger, getInterfaceAddresses, enableInet6)
+}
+
+// NetworkInterfaceAddressGetter matches the signature of net.InterfaceByName() to allow for test mocks.
+type NetworkInterfaceAddressGetter func(name string) ([]netip.Addr, error)
+
+// getFirstAddressOf returns the first IPv4/IPV6 address of the supplied interface names, omitting any link-local addresses.
+func getFirstAddressOf(names []string, logger log.Logger, interfaceAddrsFunc NetworkInterfaceAddressGetter, enableInet6 bool) (string, error) {
+	var ipAddr netip.Addr
+
+	// When passing an empty list of interface names, we select all interfaces.
+	if len(names) == 0 {
+		infs, err := net.Interfaces()
+		if err != nil {
+			return "", fmt.Errorf("failed to get interface list and no interface names supplied: %w", err)
+		}
+		ifNames := make([]string, len(infs))
+		for i, v := range infs {
+			ifNames[i] = v.Name
+		}
+		names = ifNames
+	}
+
+	level.Debug(logger).Log("msg", "looking for addresses", "inf", fmt.Sprintf("%s", names), "inet6enabled", enableInet6)
+
+	for _, name := range names {
+		addrs, err := interfaceAddrsFunc(name)
+		if err != nil {
+			level.Warn(logger).Log("msg", "error getting addresses for interface", "inf", name, "err", err)
+			continue
+		}
+
+		if len(addrs) <= 0 {
+			level.Warn(logger).Log("msg", "no addresses found for interface", "inf", name, "err", err)
+			continue
+		}
+		if ip := filterBestIP(addrs, enableInet6); ip.IsValid() {
+			// Select the best between what we've received
+			ipAddr = filterBestIP([]netip.Addr{ip, ipAddr}, enableInet6)
+		}
+		level.Debug(logger).Log("msg", "detected highest quality address", "ipAddr", ipAddr.String(), "inf", name)
+		if ipAddr.IsLinkLocalUnicast() || !ipAddr.IsValid() {
+			level.Debug(logger).Log("msg", "ignoring address", "ipAddr", ipAddr.String(), "inf", name)
+			continue
+		}
+
+		if enableInet6 {
+			if ipAddr.Is6() {
+				return ipAddr.String(), nil
+			}
+			continue
+		}
```

Comment:
Why is this code necessary? `filterBestIP()` should never return an IPv6 address if IPv6 is disabled.

Reply:
This is needed because an IPv6 address may not be present on all interfaces.  So if the best match exists on a later interface, removing this block would mean that we would get an IPv4 address.  `filterBestIP()` is only collecting from a single interface.  The good news is that we have test coverage for this case.

I suppose this speaks to the prior language of `preferInet6` vs `enableInet6`.  But in the case that we enable IPv6, and there is an address present, I feel that we should use that rather than returning an acceptable IPv4 value.  What do you think?